packages/hasher/src/lockfile.ts

```diff
@@ -0,0 +1,91 @@
+import path from "path";
+import findUp from "find-up";
+import fs from "fs-extra";
+import { parse as parseYarnLock } from "@yarnpkg/lockfile";
+import { readWantedLockfile } from "@pnpm/lockfile-file";
+import { Dependencies } from "./resolveExternalDependencies";
+import { nameAtVersion } from "./helpers";
+
+type LockDependency = {
+  version: string;
+  dependencies?: Dependencies;
+};
+
+export type ParsedLock = {
+  type: "success" | "merge" | "conflict";
+  object: {
+    [key in string]: LockDependency;
+  };
+};
+
+const memoization: { [path: string]: ParsedLock } = {};
+
+export async function parseLockFile(packageRoot: string): Promise<ParsedLock> {
+  const yarnLockPath = await findUp("yarn.lock", { cwd: packageRoot });
+
+  // First, test out whether this works for yarn
+  if (yarnLockPath) {
+    if (memoization[yarnLockPath]) {
+      return memoization[yarnLockPath];
+    }
+
+    const yarnLock = fs.readFileSync(yarnLockPath).toString();
+    const parsed = parseYarnLock(yarnLock);
+
+    memoization[yarnLockPath] = parsed;
+
+    return parsed;
+  }
+
+  // Second, test out whether this works for pnpm
+  // TODO: support rush common\config\rush\pnpm-lock.yaml
+  let pnpmLockPath = await findUp("pnpm-lock.yaml", { cwd: packageRoot });
+
+  if (pnpmLockPath) {
+    if (memoization[pnpmLockPath]) {
+      return memoization[pnpmLockPath];
+    }
+
+    const parsed = await readWantedLockfile(path.dirname(pnpmLockPath), {
+      ignoreIncompatible: true
+    });
+
+    const object: {
+      [key in string]: LockDependency;
+    } = {};
+
+    if (parsed && parsed.packages) {
+      for (const [pkgSpec, snapshot] of Object.entries(parsed.packages)) {
+        const specParts = pkgSpec.split(/\//);
+
+        const name = `${specParts[0] !== "" ? `@${specParts[0]}/` : ""}${
+          specParts[1]
+        }`;
+
+        const version = specParts[2];
```

Comment:
I'll fix the @scoped/package names bug tmw. But I don't know if I'll be able to support the tgz uris...

Reply:
Great! I don't know how important the tgz stuff is. Backfill won't do lookup on internal dependencies through the lock file since it will find a match on internal dependencies in the workspace data.

Reply:
```
const specParts = pkgSpec.split(/\//);
const name = specParts.length > 3 ? `${specParts[1]}/${specParts[2]}` : specParts[1];
const version = specParts.length > 3 ? specParts[3] : specParts[2];
```

Reply:
This is fixed & tested. The file: scheme uri is rush syntax for dealing with internal deps I think... as such, I think we'll need another release to make it work (do we hash the .tgz file as a version ,etc?)